File: VB/WpfApplication/RowStateHelper/RowStateHelper.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Windows
Imports DevExpress.Xpf.Grid
Imports System.Windows.Data

Namespace WpfApplication
	Public NotInheritable Class RowStateHelper
		Public Shared ReadOnly IsHelperEnabledProperty As DependencyProperty = DependencyProperty.RegisterAttached("IsHelperEnabled", GetType(Boolean), GetType(RowStateHelper), New UIPropertyMetadata(False, New PropertyChangedCallback(AddressOf OnIsHelperEnabledChanged)))

		Private Sub New()
		End Sub
		Public Shared Function GetIsHelperEnabled(ByVal target As DependencyObject) As Boolean
			Return CBool(target.GetValue(IsHelperEnabledProperty))
		End Function
		Public Shared Sub SetIsHelperEnabled(ByVal target As DependencyObject, ByVal value As Boolean)
			target.SetValue(IsHelperEnabledProperty, value)
		End Sub
		Private Shared Sub OnIsHelperEnabledChanged(ByVal o As DependencyObject, ByVal e As DependencyPropertyChangedEventArgs)
			OnIsHelperEnabledChanged(o, CBool(e.OldValue), CBool(e.NewValue))
		End Sub
		Private Shared Sub OnIsHelperEnabledChanged(ByVal o As DependencyObject, ByVal oldValue As Boolean, ByVal newValue As Boolean)
			Dim view As GridViewBase = TryCast(o, GridViewBase)
			RemoveHandler view.ShowingEditor, AddressOf RowStateHelper_ShowingEditor
			If newValue Then
				AddHandler view.ShowingEditor, AddressOf RowStateHelper_ShowingEditor
			End If

		End Sub

		Private Shared Sub RowStateHelper_ShowingEditor(ByVal sender As Object, ByVal e As ShowingEditorEventArgs)
			e.Cancel = Not IsRowEnabled(e.Row)
		End Sub

		Public Shared Function IsRowEnabled(ByVal row As Object) As Boolean
			Dim datarow As GridItem = TryCast(row, GridItem)
			If datarow Is Nothing Then
				Return True
			End If
			Return datarow.AllowEdit
		End Function

	End Class
End Namespace

File: CS/WpfApplication/RowStateHelper/RowStateHelper.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Windows;
using DevExpress.Xpf.Grid;
using System.Windows.Data;

namespace WpfApplication
{
    public static class RowStateHelper
    {
        public static readonly DependencyProperty IsHelperEnabledProperty = DependencyProperty.RegisterAttached("IsHelperEnabled", typeof(bool), typeof(RowStateHelper), new UIPropertyMetadata(false, new PropertyChangedCallback(OnIsHelperEnabledChanged)));

        public static bool GetIsHelperEnabled(DependencyObject target)
        {
            return (bool)target.GetValue(IsHelperEnabledProperty);
        }
        public static void SetIsHelperEnabled(DependencyObject target, bool value)
        {
            target.SetValue(IsHelperEnabledProperty, value);
        }
        private static void OnIsHelperEnabledChanged(DependencyObject o, DependencyPropertyChangedEventArgs e)
        {
            OnIsHelperEnabledChanged(o, (bool)e.OldValue, (bool)e.NewValue);
        }
        private static void OnIsHelperEnabledChanged(DependencyObject o, bool oldValue, bool newValue)
        {
            GridViewBase view = o as GridViewBase;
            view.ShowingEditor -= RowStateHelper_ShowingEditor;
            if (newValue)
                view.ShowingEditor += RowStateHelper_ShowingEditor;

        }

        static void RowStateHelper_ShowingEditor(object sender, ShowingEditorEventArgs e)
        {
            e.Cancel = !IsRowEnabled(e.Row);
        }

        public static bool IsRowEnabled(object row)
        {
            GridItem datarow = row as GridItem;
            if (datarow == null)
                return true;
            return datarow.AllowEdit;
        }

    }
}